### ISSUE TITLE
Support generics in generated types

### ISSUE THREAD

**Issue by heuels** [octokit/webhooks.js#171]
Hey guys 🙂 

First off, thanks for the awesome project! After #168 got merged I started using generated types for `workflow_dispatch` event and it works great. However having to type cast `inputs` of the event through `any` pains me a lot:

```ts
type WebhookPayloadWorkflowDispatchInputs = {};
type WebhookPayloadWorkflowDispatch = {
  inputs: WebhookPayloadWorkflowDispatchInputs;
  ref: string;
  repository: PayloadRepository;
  organization: WebhookPayloadWorkflowDispatchOrganization;
  sender: WebhookPayloadWorkflowDispatchSender;
  workflow: string;
};
```

```ts
type FooBarInput = { foo: string; bar: string };

const onWorkflowDispatch = (event: Webhooks.WebhookPayloadWorkflowDispatch) => {
  const { foo, bar } = (event.inputs as any) as FooBarInput;
  ...
}
```

However, I do believe that a better developer experience can be achieved by introducing generics in generated types:

```ts
type WebhookPayloadWorkflowDispatch<T> = {
  inputs: T;
  ref: string;
  repository: PayloadRepository;
  organization: WebhookPayloadWorkflowDispatchOrganization;
  sender: WebhookPayloadWorkflowDispatchSender;
  workflow: string;
};
```

```ts
type FooBarInput = { foo: string; bar: string };

const onWorkflowDispatch = (event: Webhooks.WebhookPayloadWorkflowDispatch<FooBarInput>) => {
  const { foo, bar } = event.inputs;
  ...
}
```

It is something that should be raised here or in [gimenete/type-writer](https://github.com/gimenete/type-writer)? I'd be happy to move this issue there if that's the case. Anyway, I would be happy to assist with a PR if/when we figure out some kind of a solution for this.

**Comment by gr2m** [comment]
I think maybe we could start defining JSON Schemas in https://github.com/octokit/webhooks, besides the examples? We could still use the examples to validate against the schema, but use the schema to generate the types for `@octokit/webhooks`.

There is an interest in JSON Schemas for webhooks from both users and GitHub, see https://github.com/github/rest-api-description/issues/21. We could accelerate that process with a community run effort in `octokit/webhooks`, similar how https://github.com/octokit/routes/ was the starting point for https://github.com/github/rest-api-description

What do you think?

**Comment by heuels** [comment]
Using JSON Schemas does seem reasonable to me. However I just looked through the "[Webhook events and payloads](https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads)" and it looks like that descriptions there are too inconsistent for us to be able to parse them. I guess we could start by composing the schema manually.

**Comment by gr2m** [comment]
> I guess we could start by composing the schema manually.

Yes, that's what I would suggest. We can bootstrap them the examples in the repository, see https://github.com/octokit/webhooks/tree/master/payload-examples/api.github.com. Unfortunately `@gimenete/type-writer` does not support a JSON Schema as far as I can tell, but if I recall correctly, I used another library to create schemas from JSON inputs. And more examples can be added easily which will help make the schema more accurate and will help with testing the schema in future.

Shall we take this discussion over to the [octokit/webhooks](https://github.com/octokit/webhooks) repository?

**Comment by gr2m** [comment]
closing in favor of https://github.com/octokit/webhooks/issues/136
